src/routes/class-base-route.php

```diff
@@ -79,80 +94,288 @@ protected function get_route_name() : string {
    * @return array Either an array of options for the endpoint, or an array of arrays for multiple methods.
    */
   protected function get_callback_arguments() : array {
-    return array(
+    return [
       'methods'  => static::READABLE,
-      'callback' => array( $this, 'route_callback' ),
-    );
+      'callback' => [ $this, 'route_callback' ],
+      'permission_callback' => [ $this, 'permission_callback' ],
+    ];
   }
 
   /**
-   * Response handler for unknown errors
+   * Verifies everything is ok with request
    *
-   * @param  array $data (Optional) data to output.
-   * @return mixed
+   * @param  \WP_REST_Request $request WP_REST_Request object.
+   * @param  string           $required_filter (Optional) Filter that needs to exist to verify this request.
+   *
+   * @throws Unverified_Request_Exception When we should abort the request for some reason.
+   *
+   * @return array            filtered request params.
    */
-  protected function rest_response_handler_unknown_error( array $data = array() ) {
+  protected function verify_request( \WP_REST_Request $request, string $required_filter = '' ): array {
```

Comment:
Why is this in the base abstract class? It would make more sense if it's in a separate class IMO. Because this way, all the classes extending the base route class will have this method available to it. And they may not use it, which kinda smells like a SRP violation to me.

Reply:
This is actually seriously useful and cuts down on repeated code A LOT. All routes should (and they do) use it as the first thing that happens in their `route_callback`. This does a number of things that were previously repeated in the code.

This method does things like:
- If the route requires a filter to do it's thing (which needs to be defined in the project) - it verifies that it exists (`has_filter()`)
- If the route has required parameters - it will check that they are all set
- If route requires authorization - it will check that the authorization parameter is added and correct for this request
- It will check captcha is correct (if used by route)

And you can easily set which of these things are used depending on some easy functions you can set per route (for example you have `get_required_params(): array` method. You can return 1 or more parameter names here and if you do, `verify_request()` will check if they're set and if not throw an exception with the proper response set (which we then catch in `route_callback` and return the response)

I'm actually really proud of this functionality, it allows you to cut back dramatically on boilerplate in routes and to dictate how the route behaves and to enable complex functionality for such route (like requiring authorization) with easy to understand methods / parameters.

Reply:
I guess I'm not used to seeing it being built like that. Regular WP rest has a security callback that gets called before the request is executed (but should be only used for authorization checks), and they usually verify things using schema.

https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#resource-schema

I won't argue about its usefulness, just looked oddly specific for a general base rout 🤷‍♂️

Reply:
Yeah I agree that schema would be useful, didn't know that existed TBH (tho not sure if you specifiy different schema for POST vs GET, which is something we need in `buckaroo-response-handler` route - as that routes has (different) required GET and POST parameters.

At this point I'm on a really tight deadline (and we have to add Mailchimp as well as a new feature request while the deadline stays the same) so I'd rather not refactor this for now.

Reply:
No, it's ok, leave it in. You have tests so you can always refactor later :trollface: